NodeJS adapters

ScottLindley opened the pull request:
This PR adds two general adapters to make it easier to work with node streams.

**`fromNodeReadable`**: Create a ts-stream ReadableStream from a Node.JS ReadableStream.

**`fromNodeWritable`**: Create a ts-stream WritableStream from a Node.JS WritableStream.

**May address the following issues**
https://github.com/poelstra/ts-stream/issues/17
https://github.com/poelstra/ts-stream/issues/40
https://github.com/poelstra/ts-stream/issues/46

ScottLindley committed:
add `fromNodeReadable` and `fromNodeWritable` adapter functions for creating ts-streams from NodeJS streams

ScottLindley committed:
update README to include NodeJS adapter functions

ScottLindley committed:
add tests for node adapters

poelstra commented on README.md:
````diff
@@ -222,6 +222,27 @@ sink.result().then(
+  .map((chunk) => String(chunk))
+  .map((letters) => letter.toUpperCase())
+  .pipe(dest);
+```
````
Nitpick: please add a newline

poelstra commented on src/lib/node.ts:
```diff
@@ -209,3 +209,75 @@ export function pipeToNodeStream<T>(
+ */
+export function fromNodeReadable(
+	nodeReadable: NodeJS.ReadableStream
+): ReadableStream<string | Buffer> {
```
What if the NodeJS stream is in object mode? It would be helpful if one could specify that as `fromNodeReadable<MyObject>(someNodeStream)`. (Defaulting to string | Buffer)

Object-mode streams are the main purpose of ts-stream, so makes sense to make working with them easy.

ScottLindley replied:
Great callout! I've updated the type signature to support an optional generic type argument.

```typescript
// creates a ts-stream of type `Stream<MyObjectType>`
const withObjectMode = fromNodeReadable<MyObjectType>(someStreamInObjectMode);

// creates a ts-stream of type `Stream<string | Buffer>` since the caller did not specify a type
const noObjectmode = fromNodeReadable(someStreamNotInObjectMode);
```

Follow-up questions: do you think this is a potential foot gun? Should I include some documentation around this? I could imagine that someone might be working with a stream in object mode but fail to specify the type for `fromNodeReadable` which means the ts-stream will work, but the type may be incorrect.

poelstra replied:
> do you think this is a potential foot gun

In a way, yes, but I think it's the best way to do it given that the Node typings don't provide any way of correctly typing object mode streams...

Including the lines you show here in the docs is a very good idea indeed!

poelstra commented on src/lib/node.ts:
````diff
@@ -209,3 +209,75 @@ export function pipeToNodeStream<T>(
+ *
+ * Usage example:
+ * ```ts
+ * const sink = toNodeWritable(fs.createWriteStream('sink.txt'));
````
Typo: `fromNodeWritable`

poelstra commented on src/lib/node.ts:
```diff
@@ -209,3 +209,75 @@ export function pipeToNodeStream<T>(
+/**
+ * Create a ts-stream WritableStream from a Node.JS WritableStream.
+ *
+ * @see pipeToNodeStream for implementation details as this is a thinn wrapper.
```
Typo: `thin`

poelstra commented on src/lib/node.ts:
```diff
@@ -209,3 +209,75 @@ export function pipeToNodeStream<T>(
+		} catch (err) {
+			await endStream(err as Error);
+		}
+	})();
```
This IIFE can cause an unhandled promise rejection exception (if `stream.end()` fails).
Not sure what to do about it though. Perhaps it makes sense to try and 'back-fire' it as an error event on the readable.

For example, if `stream.write()` fails, and subsequently `stream.end(err)` also returns an error.

ScottLindley replied:
Hmm, that's a good point. I think it's appropriate to emit an error event on the `nodeReadable` since we'll only attempt to act upon an error even on that node stream once here in this function.

ScottLindley replied:
Updated to emit an error should `steam.end` reject in the `endStream` helper above.

poelstra replied:
Yes, good solution to handle it in the helper!

poelstra commented on src/lib/node.ts:
```diff
@@ -209,3 +209,75 @@ export function pipeToNodeStream<T>(
+		if (!stream.isEndingOrEnded()) return stream.end(err);
+	};
+
+	stream.aborted().catch(endStream);
```
Does it make sense to also destroy the readable with the abort error?

ScottLindley replied:
Just to clarify, are you asking if we should emit an `error` event on behalf of the `nodeReadable` stream?

poelstra replied:
Almost, I meant to call its `destroy()`.

poelstra reviewed (changes requested):
Hey Scott,

Thanks a lot for this! I think it will make quite a lot of people happy :)

Code looks good, I have a few remarks/questions.

ScottLindley committed:
Fix comment typos and README formatting

ScottLindley committed:
handle uncaught errors, allow for overriding stream type to support `objectMode`

poelstra commented on src/lib/node.ts:
```diff
@@ -230,13 +231,25 @@ export function pipeToNodeStream<T>(
  * @returns Readable stream of type `string | Buffer`
  */
-export function fromNodeReadable(
+export function fromNodeReadable<T = void>(
```
Hmm, why not just `export function fromNodeReadable<T = Buffer | string>(...): ReadableStream<T>` ?

poelstra commented on src/lib/node.ts:
```diff
@@ -230,13 +231,25 @@ export function pipeToNodeStream<T>(
+			 * the ts-stream a maximum one time.
+			 */
+			try {
+				return stream.end(err);
```
Can you add a test for this?
Because I'm pretty sure it doesn't work right now (as the promise is directly returned, a rejection will not be caught).

ScottLindley replied:
Could you help me figure out how to test this actually? Under what conditions would `stream.end` reject? I could just override the method to force it to reject, but then the stream would never finish, right?

```ts
const stream = new Stream<string>();
stream.end = () => Promise.reject("uh oh!");

// This will just hang because the ts-stream cannot properly end... 🤔 
const result = await fromNodeReadable(someReadable).toArray();
```

poelstra replied:
Please see my remark at your latest tests. You'll need to use `.forEach()`, and instead of returning a write failure (first argument), just return a failure to the ender callback (second argument).

poelstra commented on src/lib/node.ts:
```diff
@@ -230,13 +231,25 @@ export function pipeToNodeStream<T>(
+			try {
+				return stream.end(err);
+			} catch (e) {
+				nodeReadable.emit("error", e);
```
I believe events should only ever be emitted by an object itself, not by any external entity (i.e. I think `emit()` should actually have been a protected method).
To ensure correct internal state of the object (nodeReadable), I prefer using a public API if it's available, and I think that should be the [`destroy()`](https://nodejs.org/api/stream.html#stream_readable_destroy_error) method in this case, which also accepts an optional error.

poelstra commented on src/lib/node.ts:
```diff
@@ -209,3 +209,88 @@ export function pipeToNodeStream<T>(
+			}
+			await endStream();
+		} catch (err) {
+			await endStream(err as Error);
```
Sorry, forgot to comment on this one in the first review...
It is better to ensure `err` really is an Error. Something like:

```ts
await endStream(err instanceof Error ? err : new Error(`${err}`));
```

Not the prettiest, but at least it guarantees it's an Error, and will work for strings (which I suppose is the most common non-Error thing someone would 'throw').

ScottLindley replied:
Added a helper function that's now being used in a few places to ensure we actually have an instance of an Error.

poelstra reviewed (commented):
Nice improvements!

Given the trickyness around correctly handling errors, it would be very helpful if you could add some additional tests for `fromNodeReadable`:
1. When a ts-stream's `write()` fails, it should properly destroy the nodeReadable (and ts-stream, of course)
2. Same when the ts-stream's `end()` fails
3. Same when a `write()` fails, and subsequently also the `end()` fails.

ScottLindley commented:
@poelstra just want to communicate that I'm about to step away on vacation! So just know I'm not abandoning this PR and plan to pick it back up once I've returned.

poelstra commented:
Cool, thnx for the heads up, have a good holiday!

1 jul. 2021 14:35:59 Scott Lindley ***@***.***>:

> @poelstra[https://github.com/poelstra] just want to communicate that I'm about to step away on vacation! So just know I'm not abandoning this PR and plan to pick it back up once I've returned.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub[https://github.com/poelstra/ts-stream/pull/62#issuecomment-872209665], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AAH2MXEAPIDJBCVO2UQ43SLTVROK5ANCNFSM47LQQZFQ].
>  [data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEgAAABICAYAAABV7bNHAAAAAXNSR0IArs4c6QAAAARzQklUCAgICHwIZIgAAAArSURBVHic7cEBDQAAAMKg909tDjegAAAAAAAAAAAAAAAAAAAAAAAAAAA+DFFIAAEctgHwAAAAAElFTkSuQmCC###24x24:true###][Tracking afbeelding][https://github.com/notifications/beacon/AAH2MXAZOGEUA4DDKOUVMZTTVROK5A5CNFSM47LQQZF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOGP6N2AI.gif]
>

poelstra commented on src/lib/node.ts:
```diff
@@ -231,45 +231,50 @@ type FromNodeReadableDataType<T> = T extends void ? string | Buffer : T;
+export function fromNodeReadable<T = string | Buffer>(
+	nodeReadable: NodeStream.Readable
+): ReadableStream<T> {
+	let stream = new Stream<T>();
```
Leave as `const`

ScottLindley replied:
oops!

poelstra commented on src/lib/node.ts:
```diff
@@ -231,45 +231,50 @@ type FromNodeReadableDataType<T> = T extends void ? string | Buffer : T;
+
+	stream.aborted().catch(handleError);
+	nodeReadable.once("error", handleError);
+	nodeReadable.once("close", () => endStream());
```
Can this be just `endStream` instead of the extra closure? (Not sure.)

ScottLindley replied:
It could be, but I wanted to make sure that we are not calling `endStream` with any arguments. Although, the documentation doesn't specify any function arguments for the callback https://nodejs.org/api/stream.html#stream_event_close_1

poelstra replied:
Hmyeah, makes sense, especially because endStream indeed could receive an argument. Ok, let's keep it!

poelstra commented on src/test/node.ts:
```diff
@@ -51,6 +51,44 @@ describe("node", () => {
+			const nodeStream = Readable.from(
+				(async function* () {
+					// This rejected promise will be passed to the stream's `write()` call, which will cause `write` itself to reject.
+					yield Promise.reject(errorMessage);
```
I would change this to perform 'normal' writes from the Node stream, and instead create a custom (ts-stream) sink that causes a rejected write to happen. This is trivial to do with something like:

```ts
let ended: Error | undefined;
let aborted: Error | undefined;
stream.forEach(
  (chunk) => Promise.reject(new Error("write failed")), // note the Error ;)
  (endErr) => { ended = endErr },
  (abortErr) => { aborted = abortErr; }
);
```

And then check whether the stream is correctly ended (i.e. `endError` should contain that write failed error, and it should not be aborted (I think)).

poelstra reviewed (commented):
Welcome back :-D
Thanks for adding the tests, the fix also looks good.

ScottLindley committed:
add tests for `fromNodeReadable` error handling

ScottLindley committed:
add details in README for generic type arguments

ScottLindley committed:
address callback race conditions in `fromNodeReadable` tests

ScottLindley commented:
@poelstra thanks for the quick review. I've updated the tests but I think I'm running into some race condition that I can't reproduce locally around when the `forEach` callbacks are called. My latest commit attempts to address that, but it doesn't appear to have done the trick. I'll take a look again sometime soon but in the meantime, let me know if you can spot what's going on.
I've also added a small bit in the README around using the optional generic type argument.

poelstra commented on README.md:
````diff
@@ -244,6 +244,16 @@ source
+// Produces a ts-stream of type `Stream<{ message: string; }>`.
+const b = fromNodeReadable<{ message: string; }>(someNodeStreamInObjectMode);
+```
+
````
Nice!

poelstra commented on src/test/node.ts:
```diff
@@ -121,21 +134,29 @@ describe("node", () => {
+							Promise.reject(new Error(writeErrorMessage)),
+						// simulate a failing `end` call.
+						(_endErr) => {
+							resolve();
```
I think this is definitely not the way to go, because you're now resolving a promise as soon as the ender is called, but there's a whole other chain going to happen (waiting for the ender to be resolved, then propagating that back up the chain, then leading to the final `.result()` promise being resolved, etc.).

I suppose it's best to revert this last commit, and continue from the one before this (996a2f6ff2542f2d277e4b1b690e0c27bab6ca9f), which seems to be close to how it should be.

poelstra commented on src/test/node.ts:
```diff
@@ -51,6 +51,126 @@ describe("node", () => {
+				.catch((e) => e);
+
+			expect(nodeStream.destroyed).to.equal(true);
+			expect(aborted).to.equal(undefined);
```
Add an additional assertion to verify that the "end failed" message was indeed the reason for ending things?

poelstra commented on src/test/node.ts:
```diff
@@ -51,6 +51,126 @@ describe("node", () => {
+				.catch((e) => e);
+
+			expect(nodeStream.destroyed).to.equal(true);
+			expect(aborted).to.equal(undefined);
```
Same here, please add assertions on the final errors of both streams.

poelstra reviewed (commented):
Thanks for the hard work on the tests!

The first attempts at the tests look good (i.e. the ones without the 'wrapper promise').

I can think of 2 reasons why you're seeing race conditions:
1. You're calling `tsStream.end()` without a second argument. This means that the stream's final `result()` (and thus the result of e.g. `toArray()` and `forEach()`) is going to be resolved directly after that end call resolves.
2. There are certain places in Node's stream which defer events from being emitted to `process.nextTick()`, which may cause a certain delay. I don't think this is the cause here.

I suppose the code in `endStream` suffers from point 1: it awaits `stream.end()`, and only then goes on to destroy the nodeReadable. This means that the stream's final result may resolve around the same time as that await in endStream, such that the test code continues before the catch block in endStream is triggered.

Probably the most correct way to prevent this, is by passing a second argument to `stream.end()`, which is a promise that you'll 'manually' resolve/reject with the result of the end call, but only after any other 'side effects' have been executed (`nodeReadable.destroy()`, in this case).
The code for this is a bit convoluted though. Something like (untested):

```ts
const resultDeferred = defer();
try {
  await stream.end(err, resultDeferred.promise);
  resultDeferred.resolve();
} catch (e) {
  const endError = ensureError(e);
  resultDeferred.reject(endError);
  nodeReadable.destroy(endError);
}
```

Note: the `resultDeferred.reject()` is placed before `destroy()`, in case that also throws. The result of that reject is only going to be observable in the next microtick, so it's no problem that it seems 'reversed' compared to what you'd expect.

ScottLindley commented:
@poelstra life got busy and I don't think I can make space for this anymore. I appreciate the review you've given! I'm going to close this PR but leave the branch in case someone else wants to pick it up. 👋

poelstra commented:
Sure, no worries! Thanks for the heads-up. I've cloned the branch locally and may pick it up later.
